### PR TITLE
Add explanatory text to TorchDynamo example notebook

### PR DESCRIPTION
**Pull request by ramiro050**
This commit makes the TorchDynamo example notebook self contained by adding explanations for what each code cells represents.

### CODE REVIEW

File: torchdynamo_poc/TorchDynamo Demo.ipynb [hunk 1]
```diff
@@ -85,8 +90,8 @@
     {
      "data": {
       "text/plain": [
-       "tensor([[1.0413, 1.9294, 1.4181],\n",
-       "        [1.6091, 0.9600, 1.5658]])"
+       "tensor([[2., 2., 2.],\n",
```
Comment (dellis23) [inline torchdynamo_poc/TorchDynamo Demo.ipynb:93]: Much nicer!

File: torchdynamo_poc/TorchDynamo Demo.ipynb [hunk 2]
```diff
@@ -251,54 +285,89 @@
...
+      "\u001b[0;31mTorchMlirCompilerError\u001b[0m: Lowering Torch Backend IR -> Linalg-on-Tensors Backend IR failed with the following diagnostics:\nerror: failed to legalize operation 'torch.constant.str'\nnote: see current operation: %0 = \"torch.constant.str\"() {value = \"Hello!\"} : () -> !torch.str\nerror: Module does not conform to the linalg-on-tensors backend contract. See dialect conversion legality information above.\n\n\nError can be reproduced with:\n$ torch-mlir-opt -pass-pipeline='torch-backend-to-linalg-on-tensors-backend-pipeline' /tmp/PrintStatementModule.mlir\nAdd '-mlir-print-ir-after-all -mlir-disable-threading' to get the IR dump for debugging purpose.\n"
```
Comment (dellis23) [inline torchdynamo_poc/TorchDynamo Demo.ipynb:326]: This error makes it seem like the problem is that we're using a string.  Is the error just misleading?  Would this fail with trying to print tensors too?

File: torchdynamo_poc/TorchDynamo Demo.ipynb [hunk 3]
```diff
@@ -219,7 +253,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 9,
+   "execution_count": 10,
    "id": "964ddfde",
    "metadata": {},
    "outputs": [
```
Comment (dellis23) [inline torchdynamo_poc/TorchDynamo Demo.ipynb:259]: It might be worth adding immediately after this something that demonstrates that the print statement is stripped from the trace and/or state it explicitly (e.g. "Note the lack of the print statement in the generated graph.").

File: torchdynamo_poc/TorchDynamo Demo.ipynb [hunk 4]
```diff
@@ -309,29 +378,79 @@
...
+   "id": "3280d186",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
       "Compiling graph...\n",
       "torch.fx graph:\n",
       "graph():\n",
```
Comment (dellis23) [inline torchdynamo_poc/TorchDynamo Demo.ipynb:456]: These graphs are quite hard to follow, but I don't have any good ideas on how to improve them here.  Seems like it might need a dedicated deep dive in a separate doc (though maybe that's too low level for anyone but us?).  In particular I can't tell how the sub-graphs / functions are connected.

File: torchdynamo_poc/TorchDynamo Demo.ipynb [hunk 5]
```diff
@@ -309,29 +378,79 @@
...
+   "source": [
+    "foo(torch.ones((2, 3)), torch.ones((2, 3)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "349c7526",
+   "metadata": {},
+   "source": [
+    "Notice that because the body of the `if` statement is not used, it is not compiled. If we change the inputs so that the body of the `if` statement is needed, then TorchDynamo will reuse the compiled graphs for the other parts of the function and only compile the body of the `if` statement."
```
Comment (dellis23) [inline torchdynamo_poc/TorchDynamo Demo.ipynb:441]: Am I understanding correctly that even if you pass in another set of floats with the same dtype and shape it will only re-trace if the actual values would cause the if branch to be taken?

**Review by dellis23 — APPROVED** [review]
Thanks! This illustrates the torchdynamo benefits really well.

File: torchdynamo_poc/TorchDynamo Demo.ipynb (hunk 2 above)
Comment (ramiro050) [reply torchdynamo_poc/TorchDynamo Demo.ipynb:326]: Yeah, this error is misleading. Because the string op is not lowered to something, then the `torch.constant.str` input to the op stays in the IR when reaching the linalg-on-tensors backend, leading to the error. (The current error checking in torch-mlir just looks at the first invalid op and emits the error there)

File: torchdynamo_poc/TorchDynamo Demo.ipynb (hunk 5 above)
Comment (ramiro050) [reply torchdynamo_poc/TorchDynamo Demo.ipynb:441]: That's correct. Only the path taken by the input tensors is the path compiled. If you change the input such that a new path is needed, the compiler will be run on that new path.

File: torchdynamo_poc/TorchDynamo Demo.ipynb (hunk 4 above)
Comment (ramiro050) [reply torchdynamo_poc/TorchDynamo Demo.ipynb:456]: Yeah, there is some assumed familiarity with torch.fx graphs and MLIR graphs. The graph right under this comment

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg0_1, -1), kwargs = {})
    %mul_1 : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %mul), kwargs = {})
    return mul_1
```

represents the body of the `if` statement plus the return statement:

```python
b = b * -1
return x * b
```

since this path is now needed given the inputs.

From the previous cell, the graph

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %div : [#users=1] = call_function[target=torch.ops.aten.div](args = (%arg0_1, 2), kwargs = {})
    %max_1 : [#users=1] = call_function[target=torch.ops.aten.max](args = (%arg1_1,), kwargs = {})
    %lt : [#users=1] = call_function[target=torch.ops.aten.lt](args = (%max_1, 0), kwargs = {})
    return (div, lt)
```

represents the part of the cumputation up until the `if` statement.

```python
x = a / 2
lt = b.max() < 0
return x, lt
```

and the second graph in that output represents the return statement:

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %arg0_1), kwargs = {})
    return mul
```

```python
return x * b
```

File: torchdynamo_poc/TorchDynamo Demo.ipynb (hunk 3 above)
Comment (ramiro050) [reply torchdynamo_poc/TorchDynamo Demo.ipynb:259]: Done!

File: torchdynamo_poc/TorchDynamo Demo.ipynb (hunk 4 above)
Comment (dellis23) [reply torchdynamo_poc/TorchDynamo Demo.ipynb:456]: That makes sense.  I think the piece I'm missing is how these graphs are stitched together.  Is there another graph that's not shown that takes the results of the if statement graph and decides which other graph to call, for example?
